clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -6336,6 +6336,79 @@ static void handleSYCLIntelPipeIOAttr(Sema &S, Decl *D,
   S.addSYCLIntelPipeIOAttr(D, Attr, E);
 }
 
+SYCLIntelFPGAMaxConcurrencyAttr *Sema::MergeSYCLIntelFPGAMaxConcurrencyAttr(
+    Decl *D, const SYCLIntelFPGAMaxConcurrencyAttr &A) {
+  // Check to see if there's a duplicate attribute with different values
+  // already applied to the declaration.
+  if (const auto *DeclAttr = D->getAttr<SYCLIntelFPGAMaxConcurrencyAttr>()) {
+    const auto *DeclExpr = dyn_cast<ConstantExpr>(DeclAttr->getNThreadsExpr());
+    const auto *MergeExpr = dyn_cast<ConstantExpr>(A.getNThreadsExpr());
+    if (DeclExpr && MergeExpr &&
+        DeclExpr->getResultAsAPSInt() != MergeExpr->getResultAsAPSInt()) {
+      Diag(DeclAttr->getLoc(), diag::warn_duplicate_attribute) << &A;
+      Diag(A.getLoc(), diag::note_previous_attribute);
+    }
+    return nullptr;
+  }
+  // FIXME
+  // max_concurrency and disable_component_pipelining attributes can't be
+  // applied to the same function. Upcoming patch needs to have this code
+  // added to it:
+  // if (checkAttrMutualExclusion<IntelDisableComponentPipeline>(S, D, AL))
+  //  return;
+
+  return ::new (Context)
+      SYCLIntelFPGAMaxConcurrencyAttr(Context, A, A.getNThreadsExpr());
+}
+
+void Sema::AddSYCLIntelFPGAMaxConcurrencyAttr(Decl *D,
+                                              const AttributeCommonInfo &CI,
+                                              Expr *E) {
+  if (!E->isValueDependent()) {
+    llvm::APSInt ArgVal;
+    ExprResult Res = VerifyIntegerConstantExpression(E, &ArgVal);
+    if (Res.isInvalid())
+      return;
+    E = Res.get();
+
+    // This attribute requires a strictly positive value.
+    if (ArgVal <= 0) {
+      Diag(E->getExprLoc(), diag::err_attribute_requires_positive_integer)
+          << CI << /*positive*/ 0;
+      return;
+    }
+
+    if (const auto *DeclAttr = D->getAttr<SYCLIntelFPGAMaxConcurrencyAttr>()) {
+      const auto *DeclExpr =
+          dyn_cast<ConstantExpr>(DeclAttr->getNThreadsExpr());
+      if (DeclExpr && ArgVal != DeclExpr->getResultAsAPSInt()) {
+        Diag(CI.getLoc(), diag::warn_duplicate_attribute) << CI;
+        Diag(DeclAttr->getLoc(), diag::note_previous_attribute);
+      }
+      return;
+    }
+  }
+
+  D->addAttr(::new (Context) SYCLIntelFPGAMaxConcurrencyAttr(Context, CI, E));
+}
+
+static void handleSYCLIntelFPGAMaxConcurrencyAttr(Sema &S, Decl *D,
+                                                  const ParsedAttr &A) {
+  S.CheckDeprecatedSYCLAttributeSpelling(A);
+  // FIXME
+  // max_concurrency and disable_component_pipelining attributes can't be
+  // applied to the same function. Upcoming patch needs to have this code
+  // added to it:
+  // if (checkAttrMutualExclusion<IntelDisableComponentPipeline>(S, D, AL))
+  //  return;
+
+  if (!A.checkAtLeastNumArgs(S, 1))
+    return;
+
```

Comment:
You do not need this check since this isn't a variadic argument list (I think the common attribute handler does this for you). @AaronBallman, any thought?

Reply:
Good catch, this is a bit confusing. The argument list in Attr.td says the argument is optional, but this code here says that it's not. If the argument is optional, then this code is wrong and we probably need to generate a default value for the argument. If the argument is not optional, then this code is wrong and we should remove the `/*opt*/1` in Attr.td to automate the checking.

Reply:
Failing in what way? If it was failing to issue the diagnostic without this change, that's because the definition in Attr.td makes the argument optional.

FWIW, I ran into similar confusion when working on this PR: https://github.com/intel/llvm/pull/3442

Reply:
No. It was crashing in getArgAsExpr. GetArg with null.

Reply:
With this test case:
[[intel::max_concurrency]] void foo() {}

Reply:
Ah, sure, I can imagine that. It should have never gotten there in the first place -- if the common diagnostic handler reports an error, the `handleFooAttr()` never gets called: https://github.com/intel/llvm/blob/sycl/clang/lib/Sema/SemaStmtAttr.cpp#L773

Reply:
I see. Let me see that. Thanks.

Reply:
Interesting ... The test case above and the entire LIT test max-concurrency.cpp doesn't get into that function (ProcessStmtAttr). The function handleSYCLIntelFGPAMaxConcurrencyAttr is called from ProcessDeclAttribute.

Reply:
Er, sorry, this is a decl attribute and I confusingly pointed you to the stmt attribute handler (sorry for that confusion!). I should have pointed you to: https://github.com/intel/llvm/blob/sycl/clang/lib/Sema/SemaDeclAttr.cpp#L8915

Reply:
> Good catch, this is a bit confusing. The argument list in Attr.td says the argument is optional, but this code here says that it's not. If the argument is optional, then this code is wrong and we probably need to generate a default value for the argument. If the argument is not optional, then this code is wrong and we should remove the `/*opt*/1` in Attr.td to automate the checking.

I didn't remember having put that /*opt*/1 in there. It seems to have been added by Artem? Not sure why? I sent a message to Jessica to find out is the argument to the attribute is optional or mandatory.

Reply:
This came to pulldown. This was due to @AaronBallman's recent upstream change Automate common diagnostic checking for statement  attributes. 

Follow up patch is here: https://github.com/intel/llvm/pull/3442/ - this will remove opt